clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java

```diff
@@ -244,9 +259,54 @@ public void send(String method, Object... args) throws Exception {
         }
 
         InvocationMessage invocationMessage = new InvocationMessage(method, args);
-        String message = protocol.writeMessage(invocationMessage);
-        logger.log(LogLevel.Debug, "Sending message");
-        transport.send(message);
+        sendHubMessage(invocationMessage);
+    }
+
+    public <T> CompletableFuture<T> invoke(Class<T> returnType, String method, Object... args) throws Exception {
+        InvocationMessage invocationMessage = new InvocationMessage(method, args);
+        String id = connectionState.getNextId();
+        invocationMessage.setInvocationId(id);
+        CompletableFuture<T> future = new CompletableFuture<>();
+
+        connectionState.pendingInvocations.compute(id, (k, v) -> {
+            if (v != null) {
+                // This should never happen
+                throw new IllegalStateException("Invocation Id is already used");
+            }
+
+            InvocationRequest irq = new InvocationRequest(returnType);
+            CompletableFuture<Object> pendingCall = irq.getPendingCall();
+            // forward the invocation result to the user
+            pendingCall.thenAccept((returnValue) -> {
+                // Primitive types can't be cast with the Class cast function
+                if (returnType.isPrimitive()) {
+                    future.complete((T)returnValue);
+                } else {
```

Comment:
Where do these continuations run? Inline? On a thread pool?

Reply:
Looks like they run on the same thread unless you use the Async suffix version

Reply:
So the consumer has to chose.. hmmm I don’t love that.

Reply:
Yeah, that's generally how it works. If we want, we can force it by running `.thenApplyAsync((f) -> f)` or something similar to force the future on to the thread pool (aka "fork-join pool" in Java).

Reply:
I think we need to do that (dispatch)

Reply:
@mikaelm12 FYI for your async PR